gr-blocks/lib/symbolinterleaver_xx_impl.cc

```diff
@@ -0,0 +1,178 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2022 Johannes Demel.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "symbolinterleaver_xx_impl.h"
+#include <gnuradio/io_signature.h>
+#include <type_traits>
+
+namespace gr {
+namespace blocks {
+
+template <class T>
+typename symbolinterleaver_xx<T>::sptr symbolinterleaver_xx<T>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode)
+{
+    return gnuradio::make_block_sptr<symbolinterleaver_xx_impl<T>>(
+        interleaver_indices, is_packed, interleave_mode);
+}
+
+
+/*
+ * These are explicit template declarations to make sure that all these specialized make
+ * functions are present in the compiled shared library. So far this is unnecessary for
+ * other blocks. However, the reasons are yet unknown.
+ */
+template typename symbolinterleaver_xx<uint8_t>::sptr symbolinterleaver_xx<uint8_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<gr_complex>::sptr
+symbolinterleaver_xx<gr_complex>::make(std::vector<size_t> interleaver_indices,
+                                       bool is_packed,
+                                       bool interleave_mode);
+
+template typename symbolinterleaver_xx<float>::sptr symbolinterleaver_xx<float>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<int32_t>::sptr symbolinterleaver_xx<int32_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<int16_t>::sptr symbolinterleaver_xx<int16_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+
+template <class T>
+symbolinterleaver_xx_impl<T>::symbolinterleaver_xx_impl(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode)
+    : gr::sync_block("symbolinterleaver",
+                     gr::io_signature::make(1, 1, sizeof(T)),
+                     gr::io_signature::make(1, 1, sizeof(T))),
+      d_is_packed(is_packed),
+      d_interleave_mode(interleave_mode),
+      d_unpacker(std::make_unique<gr::blocks::kernel::unpack_k_bits>(8)),
+      d_packer(std::make_unique<gr::blocks::kernel::pack_k_bits>(8)),
+      d_interleaver(std::make_unique<symbol_interleaving>(interleaver_indices))
+{
+    if (not std::is_same<T, uint8_t>::value) {
+        is_packed = false;
+        d_is_packed = false;
+    }
+    if (is_packed and d_interleaver->interleaver_length() % 8) {
+        throw std::invalid_argument(
+            fmt::format("Packed Interleaver requires 'interleaver_indices.size()' to be "
```

Comment:
I can see how a developer might appreciate the detail in this error message basically describing the test function. Could a simpler message be more broadly understood, e.g..

"Interleaver with packed byte input requires the length of interleaver_indices to be divisible by 8. <size> is not divisible by 8."